### PR TITLE
fix(2100): rank query for mysql5.7

### PR DESCRIPTION
**Pull request by parthasl**
## Context

The RANK() function is not supported in MySQL 5.7.

## Objective

Don't use the RANK() function for aggregate view raw queries in MySQL 5.7.

## References

[/v4/builds/statuses API error when using MySQL 5.7 #2100](https://github.com/screwdriver-cd/screwdriver/issues/2100)

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by wahapo** [comment]
It worked in MySQL 5.7.

However, I have one concern. If there are jobs in the pipeline that are not running yet, the view will show nothing.


I don't know if this behavior is correct. The API response is as follows, is it as expected?
`GET /v4/builds/statuses?jobIds=9&jobIds=8&numBuilds=5&offset=0`:
```
[
    {
        "jobId": 9,
        "builds": [
            {
                "id": 28,
                "jobId": 9,
                "status": "FAILED",
                "startTime": null,
                "endTime": null,
                "rank": 1
            },
            {
                "id": 29,
                "jobId": 9,
                "status": "ABORTED",
                "startTime": null,
                "endTime": null,
                "rank": 2
            }
        ]
    },
    {
        "jobId": 8,
        "builds": []
    }
]
```

I think this PR fix is fine if it behaves the same in Postgres.

**Comment by parthasl** [comment]
> It worked in MySQL 5.7.
> 
> However, I have one concern. If there are jobs in the pipeline that are not running yet, the view will show nothing.
> 
> 
> I don't know if this behavior is correct. The API response is as follows, is it as expected?
> `GET /v4/builds/statuses?jobIds=9&jobIds=8&numBuilds=5&offset=0`:
> 
> ```
> [
>     {
>         "jobId": 9,
>         "builds": [
>             {
>                 "id": 28,
>                 "jobId": 9,
>                 "status": "FAILED",
>                 "startTime": null,
>                 "endTime": null,
>                 "rank": 1
>             },
>             {
>                 "id": 29,
>                 "jobId": 9,
>                 "status": "ABORTED",
>                 "startTime": null,
>                 "endTime": null,
>                 "rank": 2
>             }
>         ]
>     },
>     {
>         "jobId": 8,
>         "builds": []
>     }
> ]
> ```
> 
> I think this PR fix is fine if it behaves the same in Postgres.

Yes, api response from postgres - 

`/v4/builds/statuses?jobIds=18654&jobIds=18653&numBuilds=5&offset=0`

```
[
   {
      "jobId":18654,
      "builds":[
         {
            "id":270334,
            "jobId":18654,
            "status":"ABORTED",
            "startTime":"2020-05-20T01:58:34.731Z",
            "endTime":"2020-05-20T01:58:53.462Z"
         },
         {
            "id":270335,
            "jobId":18654,
            "status":"SUCCESS",
            "startTime":"2020-05-20T02:00:41.980Z",
            "endTime":"2020-05-20T02:01:26.405Z"
         },
         {
            "id":270336,
            "jobId":18654,
            "status":"SUCCESS",
            "startTime":"2020-05-20T02:02:59.843Z",
            "endTime":"2020-05-20T02:03:43.398Z"
         },
         {
            "id":270338,
            "jobId":18654,
            "status":"SUCCESS",
            "startTime":"2020-05-20T02:07:54.243Z",
            "endTime":"2020-05-20T02:08:38.122Z"
         },
         {
            "id":300240,
            "jobId":18654,
            "status":"ABORTED",
            "startTime":null,
            "endTime":null
         }
      ]
   },
   {
      "jobId":18653,
      "builds":[

      ]
   }
]
```

File: lib/rawQueries.js
Comment (wahapo) [inline lib/rawQueries.js]: I am getting the error ``Unknown column 'a.id' in 'field list'``.
It seems necessary to change as follows.
`SELECT id, jobId, status, startTime, endTime`

Comment (parthasl) [reply lib/rawQueries.js]: fixed

File: lib/rawQueries.js
Comment (parthasl) [inline lib/rawQueries.js:29]: this query is fired when a build restarted. @wahapo please help check this as well.

Comment (wahapo) [reply lib/rawQueries.js:29]: No error occurred and the build started. This query fired when the Aggregate View was displayed and when the build was (re)started from page of Aggregate View.
```
[
  [
    TextRow {
      id: 1,
      jobId: 1,
      status: 'ABORTED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 2,
      jobId: 1,
      status: 'QUEUED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 3,
      jobId: 1,
      status: 'QUEUED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 4,
      jobId: 1,
      status: 'QUEUED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 5,
      jobId: 1,
      status: 'ABORTED',
      startTime: null,
      endTime: null
    }
  ],
  [
    TextRow {
      id: 1,
      jobId: 1,
      status: 'ABORTED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 2,
      jobId: 1,
      status: 'QUEUED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 3,
      jobId: 1,
      status: 'QUEUED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 4,
      jobId: 1,
      status: 'QUEUED',
      startTime: null,
      endTime: null
    },
    TextRow {
      id: 5,
      jobId: 1,
      status: 'ABORTED',
      startTime: null,
      endTime: null
    }
  ]
]
```

Comment (parthasl) [reply lib/rawQueries.js:29]: thnx, then good to merge once approved.

**Review by wahapo — APPROVED** [review]
LGTM

**Review by vspeaks — APPROVED** [review]

**Review by vnugopal — APPROVED** [review]

**Review by kumada626 — APPROVED** [review]

File: lib/rawQueries.js
Comment (DekusDenial) [inline lib/rawQueries.js]: The postgresql one is ranking by `id DESC`, which mean the latest build comes first in ranking. Should this be `a.id <= b.id` instead so the latest build from `a` will have the rank of `1` cuz it should have the least count and u can get rid of the `+1` on the count?

Comment (parthasl) [reply lib/rawQueries.js]: did this based on current api output in postgres - https://github.com/screwdriver-cd/models/pull/456#issuecomment-638590892

Comment (parthasl) [reply lib/rawQueries.js]: you are correct, updated query

**Comment by parthasl** [comment]
[postgresql_output.log](https://github.com/screwdriver-cd/models/files/4737750/postgresql_output.log)
[mysql_output.log](https://github.com/screwdriver-cd/models/files/4737751/mysql_output.log)

**Pull request merged**
